### DISCUSSION TITLE
TypeError: fetch failed when building the app

### DISCUSSION THREAD

**Discussion by vitormarkis** [vercel/next.js#53323] [Help]
### Summary

I'm using Planetscale database proxy to runs the database `pscale connect <db-name> <branch> --port 3309`

```.env
DATABASE_URL="mysql://root@localhost:3309/db-name"
```

It's working on dev mode, also ``npx prisma studio` connects sucessfully, but when i try to build the app, in the pages where i fetch data using next async components, it throws this error for multiple pages:

### Example where I fetch data in async component
```tsx
export async function PostsSection({}: { session?: Session }) {
  const session = await getSession(headers().get("cookie") ?? "")
  const response = await fetch(`${api_endpoint}/api/posts`, {
    next: { revalidate: 10 },
  })
  const unparsedData = await response.json()
  const posts = z.array(postFeedSchema).parse(unparsedData)
// ...
```

### The error on building logs

```
...
[=== ] - info Generating static pages (7/8)(node:1591) ExperimentalWarning: The Fetch API is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
node:internal/deps/undici/undici:11118
    Error.captureStackTrace(err, this);
          ^

TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11118:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  cause: Error: connect ECONNREFUSED 127.0.0.1:3000
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1300:16)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -111,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '127.0.0.1',
    port: 3000
  }
}

Node.js v18.12.1
...
```

I assume its a error related to my PC, because the current commit on production works, but building locally it throws this error.

**Comment by mandorakannu** [comment]
This error occurs because localhost is not defined on build time and your application fetch data from localhost.

**Reply by aquiseb** [reply]
Can you please elaborate? How should we fetch /api inside the app in order to not encounter that error?

**Reply by mandorakannu** [reply]
dont use fetch API ! You need to use that function which you are written in `route.ts` file in your server component directly and if you want to cache the result then, use
```
import { cache } from "react";

cache(/*Your function*/)

// other content
```

checkout this code: [my code](https://github.com/mandorakannu/my-blogs/blob/master/app/(common)/Blogs/page.tsx)

**Comment by devjiwonchoi** [comment]
Assuming your `api_endpoint` on development is `localhost (127.0.0.1)`, the "API route" of your localhost cannot be established during the build time.

### Work Around

> Which could be a bit annoying...

#### Option 1

1.  Mock your `route.ts` with a simple express app
2. Run the express server during build time

#### Option 2

1. Deploy your Next.js app with your `route.ts`, without any dependant pages.
2. Redeploy with your pages

**Note that `route.ts` are currently not meant to be an `API Server`, but rather a `route handler`.**
